### PR TITLE
[6.0.0] Header.checkPow method

### PR DESCRIPTION
**Pull request by kushti**
In this PR, a new method added to check Ergo header's proof-of-work
```
{
     val h = CONTEXT.headers(0)
      h.checkPow
 }
```

### CODE REVIEW

**Comment by kushti** [comment]
JS test was failing due to #1023 , fixed now

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: data/shared/src/main/scala/sigma/ast/values.scala [hunk 1]
```diff
@@ -1313,6 +1313,9 @@ case class MethodCall(
     addCost(MethodCall.costKind) // MethodCall overhead
     method.costKind match {
       case fixed: FixedCost =>
+        if (method.sinceVersion > 0 && E.context.currentErgoTreeVersion < method.sinceVersion) {
+          syntax.error(s"Method ${method.name} is not supported in tree version ${E.context.currentErgoTreeVersion}")
+        }
```
[the remark this answers is not in the document]
Comment (kushti) [reply data/shared/src/main/scala/sigma/ast/values.scala:1318]: Possible, however, there are some questions to resolve then :

* NewFeature implementation by  you was expecting this: 

```
    override def isSupportedIn(vc: VersionContext): Boolean =
      vc.activatedVersion >= sinceVersion && vc.ergoTreeVersion >= sinceVersion
```
why ?

* what does tree version mean at all ? How tree of version v+1 is different from tree version v?

File: data/shared/src/main/scala/sigma/pow/Autolykos2PowValidation.scala [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+  /**
+    * Generate element of Autolykos equation.
+    */
+  private def genElementV2(indexBytes: Array[Byte], heightBytes: => Array[Byte]): BigInt = {
+    // Autolykos v. 2: H(j|h|M) (line 5 from the Algo 2 of the spec)
+    toBigInt(hash(Bytes.concat(indexBytes, heightBytes, M)).drop(1))
+  }
+
+  def hitForVersion2ForMessage(k: Int, msg: Array[Byte], nonce: Array[Byte], h: Array[Byte], N: Int): BigInt = {
```
[the remark this answers is not in the document]
Comment (kushti) [reply data/shared/src/main/scala/sigma/pow/Autolykos2PowValidation.scala:108]: done

File: data/shared/src/main/scala/sigma/pow/Autolykos2PowValidation.scala [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  */
+object Autolykos2PowValidation {
+
+  type Height = Int
+
+  val k = 32
+
+  val NStart = 26
+
+  val group: BcDlogGroup = CryptoConstants.dlogGroup
```
[the remark this answers is not in the document]
Comment (kushti) [reply data/shared/src/main/scala/sigma/pow/Autolykos2PowValidation.scala:21]: done

File: data/shared/src/main/scala/sigma/ast/values.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (kushti) [reply data/shared/src/main/scala/sigma/ast/values.scala:1318]: Rollback versioning by tree and changed NewFeature definition. The question about versioning semantics still remains.

**Comment by kushti** [comment]
@aslesarenko comments addressed, please make another pass

File: data/shared/src/main/scala/sigma/ast/values.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply data/shared/src/main/scala/sigma/ast/values.scala]: This "vc.activatedVersion >= sinceVersion && vc.ergoTreeVersion >= sinceVersion" was an intermediate version, which was fixed in the in my original more recent PRs. 
The tree version is more about bytecode format, see for example switch from v0, to v1, there size has become required. However, there is also a relation between tree version and block version, s.t. treeVersion <= activatedScriptVersion.

**Review by aslesarenko — APPROVED** [review]
LGTM, also see comment about tree version.

**Pull request merged**